server/apps/migration/core-data-jpa-to-pg/src/main/java/org/apache/james/JpaToPgCoreDataMigration.java

```diff
@@ -0,0 +1,455 @@
+            bind(JPAUsersDAO.class).in(Scopes.SINGLETON);
+        }
+
+        @ProvidesIntoSet
+        InitializationOperation configureDomainList(DomainListConfiguration configuration, JPADomainList jpaDomainList) {
+            return InitilizationOperationBuilder
+                    .forClass(JPADomainList.class)
+                    .init(() -> jpaDomainList.configure(configuration));
+        }
+
+        @ProvidesIntoSet
+        InitializationOperation configureJpaUsers(
+                EntityManagerFactory entityManagerFactory,
+                ConfigurationProvider configurationProvider,
+                JPAUsersDAO usersDAO
+        ) {
+            return InitilizationOperationBuilder
+                    .forClass(JPAUsersDAO.class)
+                    .init(() -> {
+                        usersDAO.configure(configurationProvider.getConfiguration("usersrepository", LoggingLevel.INFO));
+                        usersDAO.setEntityManagerFactory(entityManagerFactory);
+                        usersDAO.init();
+                    });
+        }
+    }
+
+    static class CoreEntityValidatorsModule extends AbstractModule {
+        @Override
+        protected void configure() {
+            Multibinder.newSetBinder(binder(), UserEntityValidator.class).addBinding().to(DefaultUserEntityValidator.class);
+            Multibinder.newSetBinder(binder(), UserEntityValidator.class).addBinding().to(RecipientRewriteTableUserEntityValidator.class);
+        }
+
+        @Provides
+        @Singleton
+        UserEntityValidator userEntityValidator(Set<UserEntityValidator> validatorSet) {
+            return new AggregateUserEntityValidator(validatorSet);
+        }
+    }
+
+    static class PostgresMailRepositoryUrlStoreModule extends AbstractModule {
+        @Override
+        protected void configure() {
+            bind(PostgresMailRepositoryUrlStore.class).in(Scopes.SINGLETON);
+            bind(MailRepositoryUrlStore.class).to(PostgresMailRepositoryUrlStore.class);
+            Multibinder.newSetBinder(binder(), PostgresModule.class)
+                    .addBinding().toInstance(org.apache.james.mailrepository.postgres.PostgresMailRepositoryModule.MODULE);
+        }
+    }
+
+    Injector getInjector() {
+        return injector;
+    }
+
+    public JpaToPgCoreDataMigration(Module module) {
+        this.injector = Guice.createInjector(module);
+        injector.getInstance(InitializationOperations.class).initModules();
+    }
+
+    void start() {
+        injector.getInstance(DomainMigration.class).doMigration();
+        injector.getInstance(UsersMigration.class).doMigration();
+        injector.getInstance(RRTMigration.class).doMigration();
+        injector.getInstance(MailRepositoryUrlMigration.class).doMigration();
+        injector.getInstance(DropListMigration.class).doMigration();
+        injector.getInstance(SieveMigration.class).doMigration();
+    }
```

Comment:
:-P Maybe that's me over-engineering but I would have been writing a `MigrationAggregate implementing the `Migration` interface and using a `Migration::and` reduction. I would have had also used a set multi binder to inject the migration steps.

Reply:
I guess that's why you wanted to have an interface.
As far as I can tell this list is exhaustive it also has the benefit of being very explicit on what the tool actually migrates I don't think additional architecture is needed here.

Reply:
Agreed